### DISCUSSION TITLE
Cross-platform support for winget-dsc

### DISCUSSION THREAD

**Discussion by ryfu-msft** [microsoft/winget-dsc#86] [General]
Starting a discussion to talk about cross-platform support as we continue to grow the functionality of our DSC resources.

**Comment by Gijsreyn** [comment]
> [!NOTE]  
> This discussion references the idea on [Support cross-platform testing](https://github.com/microsoft/winget-dsc/issues/64)

Thanks @ryfu-msft  for opening up the discussion on cross-platform support. Let me try to give it a shot.

## Context 

Even that `winget.exe` primary focus is on Windows, the DSC resources inside this repository _can_ work on cross-platform (expect the ones labelled with _Windows_. To guarantee the quality and working of these DSC resources on other platforms, it would be beneficial to have a testing strategy available. This will also help in identifying gaps that require additional developments. 

## Proposed solution

The following bullet points outline the proposed solution to help a broader community easily develop and test new and existing DSC resources across multiple platforms:

- Create a test strategy matrix in Azure Pipelines targetting Ubuntu, MacOs and Windows
- Develop unit test(s) that target specific platforms it requires testing on (`$IsWindows`, `$IsMacOs`, `$IsLinux`)
- Include `dsc.exe` to execute DSC resources
- Seperate bootstrapping of required software in each test, but keep global tool installation e.g. `PSDesiredStateConfiguration` and `dsc.exe` into a single step

## Out of scope

While this doesn't directly related to cross-platform capabilities, I thought of writing the following points down as it relates to testing:

- Standardize on documentation and help in generating it (relates in the discussion on [DSC](https://github.com/PowerShell/DSC/discussions/508))
- Module quality checks (e.g. test for DSCCapabilities, Tags, IconUri, LinkUri)
